autogen/agentchat/conversable_agent.py

```diff
@@ -300,23 +302,38 @@ def _append_oai_message(self, message: Union[Dict, str], role, conversation_id:
         """
         message = self._message_to_dict(message)
         # create oai message to be appended to the oai conversation that can be passed to oai directly.
-        oai_message = {k: message[k] for k in ("content", "function_call", "name", "context") if k in message}
+        oai_message = {
+            k: message[k]
+            for k in ("content", "function_call", "tool_calls", "tool_call_id", "name", "context")
+            if k in message and message[k] is not None
+        }
         if "content" not in oai_message:
-            if "function_call" in oai_message:
+            if "function_call" in oai_message or "tool_calls" in oai_message:
                 oai_message["content"] = None  # if only function_call is provided, content will be set to None.
             else:
                 return False
 
-        oai_message["role"] = "function" if message.get("role") == "function" else role
-        if "function_call" in oai_message:
+        for tool_call in oai_message.get("tool_calls", []):
+            function = tool_call.get("function", {})
+            if function.get("name", False):
+                function["name"] = self._normalize_name(function["name"])
+
+        if message.get("role") in ["function", "tool"]:
+            oai_message["role"] = message.get("role")
+        else:
+            oai_message["role"] = role
+
+        if oai_message.get("function_call", False) or oai_message.get("tool_calls", False):
             oai_message["role"] = "assistant"  # only messages with role 'assistant' can have a function call.
-            oai_message["function_call"] = dict(oai_message["function_call"])
         self._oai_messages[conversation_id].append(oai_message)
         return True
 
+    def _normalize_name(self, name):
+        return re.sub(r"[^a-zA-Z0-9_-]", "_", name)[:64]
+
     def send(
         self,
-        message: Union[Dict, str],
+        message: Union[List[Union[Dict, str]], Dict, str],
```

Comment:
This is a big change which deserves more careful checks. Could you please discuss the necessity of this change? My understanding is that the reply from `generate_reply` can now be a list of messages. If so, all the related method signatures will need to change.
An alternative way with less code change is to keep message as a dict, but use a reserved field to store the list when relevant, and process it inside the `generate_oai_reply` methods for LLM inference.

Reply:
yeah, but I think we can plumb it through _append_oai_message instead and that should help limit the impact of the custom message dict.

Reply:
That makes sense. We also need to check all the reply functions which assume `messages[-1]` is the received message.

Reply:
oof, yeah `messages[-1]` is not a good assumption in most places its used, `unseen_messages` or `messages_since_last_send` or `last_received_messages` would be good but that is a whole other set of changes. But if we parse `tool_responses` in generate_oai_reply then anything not explicitly looking for `tool_responses` it will see an empty message which may be worse than just reading the last of multiple tool responses.

Trying to find the shortest path toward wrapping this pr. But this plus some other nits I've been holding off on are making me want to heavily refactor or just rewrite conversable_agent :-) (dict/Type all messages, role customization, always name agents, change hooks to promisio to better handle sync/async function dichotomy,...)

Reply:
I suppose can just do
```
{
  "tool_responses": [{...},...]
  "content: "\n".join([response[content] for response in tool_responses])
}
```

Reply:
Perhaps it is better to quickly wrap up this PR using a reserved field or something that involves local changes, as it is blocking other PRs related to function/tool calls.